cmd/health_checker/main.go

```
@@ -0,0 +1,385 @@
+	flag.String("status-codes", "200", "valid response status codes")
+	flag.Bool("skip-verify", false, "skip certifiate validation")
+	flag.Int("tries", 5, "number of tries")
+	flag.Int("backoff", 1, "backoff in seconds")
+	flag.String("log-env", "development", "logging config: development or production")
+	flag.String("log-level", "error", "log level: debug, info, warn, error, dpanic, panic, or fatal")
+
+	flag.Parse(os.Args[1:])
+
+	v := viper.New()
+	v.BindPFlags(flag)
+	v.SetEnvPrefix("HEALTHCHECKER")
+	v.SetEnvKeyReplacer(strings.NewReplacer("-", "_"))
+	v.AutomaticEnv()
+
+	log, err := createLogger(v.GetString("log-env"), v.GetString("log-level"))
+	if err != nil {
+		log.Fatal(err.Error())
+	}
+
+	err = checkConfig(v)
+	if err != nil {
+		switch e := err.(type) {
+		case *errInvalidPath:
+			log.Fatal(e.Error(), zap.String("path", e.Path))
+		case *errInvalidScheme:
+			log.Fatal(e.Error(), zap.String("scheme", e.Scheme))
+		case *errInvalidStatusCode:
+			log.Fatal(e.Error(), zap.Int("code", e.Code))
+		}
+		log.Fatal(err.Error())
+	}
+
+	verbose := v.GetBool("verbose")
+	schemes := strings.Split(strings.TrimSpace(v.GetString("schemes")), ",")
+	hosts := strings.Split(strings.TrimSpace(v.GetString("hosts")), ",")
+	paths := strings.Split(strings.TrimSpace(v.GetString("paths")), ",")
+	statusCodes, _ := stringSliceToIntSlice(strings.Split(strings.TrimSpace(v.GetString("status-codes")), ","))
+
+	httpClient, err := createHTTPClient(v, log)
+	if err != nil {
+		log.Fatal(errors.Wrap(err, "error creating http client").Error())
+	}
+
+	maxTries := v.GetInt("tries")
+	backoff := v.GetInt("backoff")
+
+	for _, scheme := range schemes {
+		for _, host := range hosts {
+			for _, path := range paths {
+				url := scheme + "://" + host + path
+				if verbose {
+					log.Info("checking url", zap.String("url", url))
+				}
+				err := checkURL(httpClient, url, statusCodes, maxTries, backoff, log)
+				if err != nil {
+					log.Fatal(err.Error())
```

I think it would be better to capture any errors returned by `checkURL` and continue running the rest of the tests, printing the failures and exiting with the right status only after running all of the checks. Otherwise, an early check can fail and prevent any other checks. This could hide other failures that might help diagnose a potential issue.

---

made change as requested an added `exit-on-error` flag, which means all checks are run unless flag is set.